### DISCUSSION TITLE
basic authorization in crome

### DISCUSSION THREAD

**Discussion by mr-anton-t** [webdriverio/webdriverio#6670] [Q&A]
Hi 
I can't open the site with basic authorization in any way. 
I found a solution for chrome and added it. but it didn't help:
https://stackoverflow.com/questions/57192562/passing-basic-auth-credentials-in-google-chrome-shows-the-pop-again

conf file: 
```
services: ['chromedriver'],
    maxInstances: 1,
    logOutput: './tests-wdio/currency/logs',
    capabilities: [
        {
            browserName: "chrome",
            maxInstances: 1,
            "goog:chromeOptions": {
                args: [//'--headless',
                    '--disable-software-rasterizer',
                    '--disable-blink-features="BlockCredentialedSubresources"',
                    '--no-proxy-server',
                    '--disable-gpu',
                    '--no-sandbox',
                    '--ignore-certificate-errors']
            }
        }]
```

log:
```
[0-0] 2021-04-07T15:03:48.075Z INFO webdriver: Initiate new session using the WebDriver protocol
[0-0] 2021-04-07T15:03:48.076Z INFO webdriver: [POST] http://localhost:9515/session
[0-0] 2021-04-07T15:03:48.076Z INFO webdriver: DATA {
  capabilities: {
    alwaysMatch: { browserName: 'chrome', 'goog:chromeOptions': [Object] },
    firstMatch: [ {} ]
  },
  desiredCapabilities: { browserName: 'chrome', 'goog:chromeOptions': { args: [Array] } }
}
[0-0] DEPRECATION: An asynchronous before/it/after function took a done callback but also returned a promise. This is not supported and will stop working in the future. Either remove the done callback (recommended) or change the function to not return a promise.
[0-0] 2021-04-07T15:03:49.576Z INFO webdriver: COMMAND navigateTo("https://login:pass@test.local")
[0-0] 2021-04-07T15:03:49.576Z INFO webdriver: [POST] http://localhost:9515/session/8ebf0bddf3160fec8dcbc9576fa88d0a/url
```
I see  window with basic authorization  on the screen. and I can't go to the site

Versions from package json:
 "@wdio/cli": "7.2.3",
"chromedriver": "89.0.0",
 "wdio-chromedriver-service": "7.0.0"

**Comment by portyu9** [comment]
Hi @mr-anton-t 

The Chrome flag workaround is no longer the approach I would use.

Current WebdriverIO supports HTTP Basic Authentication directly when navigating:

```js
await browser.url('https://test.local', {
    auth: {
        user: 'login',
        pass: 'pass'
    }
})
```

So instead of:

```js
await browser.url('https://login:pass@test.local')
```

and:

```text
--disable-blink-features="BlockCredentialedSubresources"
```

use the `auth` option.

A browser Basic Authentication prompt is browser UI, not normal page DOM, so trying to automate the username/password controls as regular WebdriverIO elements is also not the right approach.

Your original versions are quite old:

```text
@wdio/cli 7.2.3
chromedriver 89
```

so the `browser.url(..., { auth })` example above is the **current WebdriverIO API**. I would upgrade WebdriverIO/Chrome first rather than trying to preserve the Chrome 89 workaround.

There is also a separate warning in your log:

```text
An asynchronous before/it/after function took a done callback
but also returned a promise.
```

That is unrelated to Basic Auth, but it should also be fixed. If your hook/test is `async`, remove the `done` callback:

```js
before(async function () {
    // await ...
})
```

rather than mixing `async`/Promises with:

```js
before(async function (done) {
    // ...
})
```

So for a current setup, the supported path is to supply the credentials through WebdriverIO's navigation `auth` option and remove the obsolete Chrome flag.

Please mark this answer as accepted if it helped, thank you.

**Reply by mr-anton-t** [reply]
5 years ago, this version of Chrome was current 🤷‍♂️

**Reply by portyu9** [reply]
This situation can still occur today if the site uses HTTP Basic Authentication.

With current WebdriverIO, you no longer need the Chrome flag or credentials embedded in the URL. Use the built-in `auth` option:

```js
await browser.url('https://test.local', {
    auth: {
        user: 'login',
        pass: 'pass'
    }
})
```

This supplies the Basic Auth credentials during navigation and avoids having to interact with the browser's authentication prompt.

So the current solution is `browser.url(..., { auth })`; the old `BlockCredentialedSubresources` Chrome workaround is no longer necessary.

Please mark this answer as accepted if it helped, thank you.
